File: vm/tests/resources/cratonvm/ChmFirstInsertRace.java
package cratonvm;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Fixture for {@code vm/tests/chm_first_insert_race.rs}: several threads make
 * the FIRST inserts into one fresh {@code ConcurrentHashMap} at once.
 *
 * <p>The default constructor defers the map's storage to its first insert, and
 * that install used to be a plain store: two first inserts each built storage,
 * the later one won, and the earlier one's key -- already reserved in the losing
 * storage -- was lost, with {@code computeIfAbsent} returning null. This is the
 * shape of Tomcat's {@code TimeBucketCounterBase.increment}, which threw
 * NullPointerException on a client's first request.
 *
 * <p>Compile: {@code javac -d vm/tests/resources
 * vm/tests/resources/cratonvm/ChmFirstInsertRace.java}
 */
public final class ChmFirstInsertRace {

    /**
     * {@code rounds} fresh maps, {@code threads} threads each doing
     * {@code computeIfAbsent} on its own key, released together. Answers the
     * number of rounds that went wrong: a null return, or a map left with fewer
     * than {@code threads} entries. The JDK answers 0.
     */
    public static int badRounds(int rounds, int threads) throws InterruptedException {
        int bad = 0;
        for (int r = 0; r < rounds; r++) {
            final ConcurrentHashMap<String, AtomicInteger> map = new ConcurrentHashMap<>();
            final CountDownLatch go = new CountDownLatch(1);
            final AtomicInteger nulls = new AtomicInteger();
            Thread[] ts = new Thread[threads];
            for (int t = 0; t < threads; t++) {
                final String key = "k" + t;
                ts[t] = new Thread(new Runnable() {
                    @Override
                    public void run() {
                        try {
                            go.await();
                        } catch (InterruptedException e) {
                            return;
                        }
                        AtomicInteger ai = map.computeIfAbsent(key, ChmFirstInsertRace::newCounter);
                        if (ai == null) {
                            nulls.incrementAndGet();
                        } else {
                            ai.incrementAndGet();
                        }
                    }
                });
                ts[t].start();
            }
            go.countDown();
            for (Thread t : ts) {
                t.join();
            }
            if (nulls.get() != 0 || map.size() != threads) {
                bad++;
            }
        }
        return bad;
    }

    private static AtomicInteger newCounter(String ignored) {
        return new AtomicInteger();
    }

    /** Same race through {@code put}: rounds whose map lost an entry. */
    public static int putBadRounds(int rounds, int threads) throws InterruptedException {
        int bad = 0;
        for (int r = 0; r < rounds; r++) {
            final ConcurrentHashMap<String, Integer> map = new ConcurrentHashMap<>();
            final CountDownLatch go = new CountDownLatch(1);
            Thread[] ts = new Thread[threads];
            for (int t = 0; t < threads; t++) {
                final String key = "k" + t;
                final Integer value = Integer.valueOf(t);
                ts[t] = new Thread(new Runnable() {
                    @Override
                    public void run() {
                        try {
                            go.await();
                        } catch (InterruptedException e) {
                            return;
                        }
                        map.put(key, value);
                    }
                });
                ts[t].start();
            }
            go.countDown();
            for (Thread t : ts) {
                t.join();
            }
            if (map.size() != threads) {
                bad++;
            }
        }
        return bad;
    }

    public static void main(String[] args) throws InterruptedException {
        int rounds = args.length > 0 ? Integer.parseInt(args[0]) : 2000;
        long t0 = System.nanoTime();
        int cia = badRounds(rounds, 4);
        int put = putBadRounds(rounds, 4);
        System.out.println("computeIfAbsent bad rounds=" + cia + "/" + rounds + ", put bad rounds=" + put + "/"
                + rounds + " (" + (System.nanoTime() - t0) / 1_000_000 + " ms)");
    }
}

File: vm/tests/chm_first_insert_race.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2024-2026 Craton Software Company

//! Concurrent FIRST inserts into one fresh `ConcurrentHashMap` must not lose a
//! key.
//!
//! `new ConcurrentHashMap<>()` defers its native storage to the first insert,
//! and `chm_segment_for_mut` installed it with a plain store. Two threads
//! inserting first each built storage and the later store won; the other
//! thread's key, already reserved in the losing storage, vanished, and
//! `computeIfAbsent` returned null. MEASURED 2026-09-12 with this fixture on
//! the release binary: 1070 of 2000 rounds bad for `computeIfAbsent`, 808 of
//! 2000 for `put`; HotSpot 0 and 0. It surfaced as Tomcat's
//! `TestRateLimitFilter` `expected:<200> but was:<0>`, a client thread killed
//! by an NPE in `TimeBucketCounterBase.increment` on its first request.

use cratonvm_vm::config::VmConfig;
use cratonvm_vm::types::Value;
use cratonvm_vm::vm::Vm;

const CLASS: &str = "cratonvm/ChmFirstInsertRace";

fn test_resources_dir() -> String {
    if let Some(generated) = option_env!("CRATONVM_TEST_CLASSES_DIR") {
        if std::path::Path::new(generated).is_dir() {
            return generated.to_owned();
        }
    }
    format!("{}/tests/resources", env!("CARGO_MANIFEST_DIR"))
}

/// The fixture starts threads and waits on a `CountDownLatch`, which needs a
/// class library: `VmConfig::default()` is synthetic-JDK mode, and without the
/// `synthetic-jdk` Cargo feature that mode has none, so `Thread.<init>` throws
/// before the race is ever reached. Same gate as `interpreter_tests.rs`. Run:
///
/// ```text
/// cargo test --release -p cratonvm-vm --features synthetic-jdk --test chm_first_insert_race
/// ```
fn library_available(test: &str) -> bool {
    if !cratonvm_vm::config::SYNTHETIC_JDK_COMPILED_IN {
        eprintln!(
            "Skipping {test}: the fixture needs java.lang.Thread; rebuild with \
             `--features synthetic-jdk`"
        );
        return false;
    }
    true
}

fn bad_rounds(method: &str, rounds: i32) -> i32 {
    let mut vm = Vm::new(VmConfig::new().with_classpath(vec![test_resources_dir()]));
    match vm.invoke(CLASS, method, "(II)I", &[Value::Int(rounds), Value::Int(4)]) {
        Ok(Some(Value::Int(n))) => n,
        other => panic!("{CLASS}.{method} did not return an int: {other:?}"),
    }
}

#[test]
fn concurrent_first_compute_if_absent_never_returns_null_or_loses_a_key() {
    if !library_available("concurrent_first_compute_if_absent_never_returns_null_or_loses_a_key") {
        return;
    }
    const ROUNDS: i32 = 300;
    let bad = bad_rounds("badRounds", ROUNDS);
    assert_eq!(
        bad, 0,
        "{bad} of {ROUNDS} rounds of four concurrent first computeIfAbsent calls on a fresh \
         ConcurrentHashMap returned null or lost a key -- the lazy segment install is racing again"
    );
}

#[test]
fn concurrent_first_put_never_loses_an_entry() {
    if !library_available("concurrent_first_put_never_loses_an_entry") {
        return;
    }
    const ROUNDS: i32 = 300;
    let bad = bad_rounds("putBadRounds", ROUNDS);
    assert_eq!(
        bad, 0,
        "{bad} of {ROUNDS} rounds of four concurrent first puts on a fresh ConcurrentHashMap lost \
         an entry -- the lazy segment install is racing again"
    );
}
